rust/src/bin/storage_stats.rs

```diff
@@ -0,0 +1,105 @@
+// Copyright (c) 2025 Sonic Operations Ltd
+//
+// Use of this software is governed by the Business Source License included
+// in the LICENSE file and at soniclabs.com/bsl11.
+//
+// Change Date: 2028-4-16
+//
+// On the date above, in accordance with the Business Source License, use of
+// this software will be governed by the GNU Lesser General Public License v3.
+
+//! A command line tool to print storage statistics for a Carmen database.
+//! The tool reads a storage path containing a Carmen database a prints various
+//! statistics about the storage using the specified output formatters.
+
+use std::{
+    io::stdout,
+    path::{Path, PathBuf},
+    sync::Arc,
+};
+
+use carmen_rust::{
+    database::{
+        self,
+        verkle::variants::managed::{
+            FullLeafNode, InnerNode, SparseLeafNode, VerkleNode, VerkleNodeFileStorageManager,
+        },
+        visitor::AcceptVisitor,
+    },
+    node_manager::cached_node_manager::CachedNodeManager,
+    statistics::{
+        PrintStatistic,
+        formatters::{
+            StatisticsFormatter, csv_writer::CSVWriter, human_readable_writer::HumanReadableWriter,
+        },
+        node_count::NodeCountVisitor,
+    },
+    storage::{
+        Storage,
+        file::{NoSeekFile, NodeFileStorage},
+    },
+};
+use clap::{Parser, ValueEnum};
+
+/// An enum representing the available output formatters. It must match the available
+/// implementations of [`StatisticsFormatter`].
+#[derive(Copy, Clone, Debug, PartialEq, Eq, ValueEnum)]
+enum Formatter {
+    Stdout,
+    Csv,
+}
+
+impl Formatter {
+    /// Instantiate the formatter corresponding to this enum variant.
+    fn to_formatter(self) -> Box<dyn StatisticsFormatter> {
+        match self {
+            Formatter::Stdout => Box::new(HumanReadableWriter::new(stdout())),
+            Formatter::Csv => Box::new(CSVWriter {}),
+        }
+    }
+}
+
+/// Storage Statistics Tool
+#[derive(Parser, Debug)]
+#[command(version, about, long_about = None)]
+struct Args {
+    /// Path to the DB directory
+    #[arg(short, long)]
+    db_path: PathBuf,
+
+    /// Output format(s) to use
+    #[arg(short, long, value_enum, num_args = 1.., default_value = "stdout")]
+    formatter: Vec<Formatter>,
+}
+
+fn main() {
+    type FileStorage = VerkleNodeFileStorageManager<
+        NodeFileStorage<InnerNode, NoSeekFile>,
+        NodeFileStorage<SparseLeafNode<2>, NoSeekFile>,
+        NodeFileStorage<FullLeafNode, NoSeekFile>,
+    >;
+
+    let args = Args::parse();
+
+    let storage_path = Path::new(&args.db_path);
+    let is_pinned = |_n: &VerkleNode| false; // We don't care about the pinned status for stats
+    let storage = FileStorage::open(storage_path).unwrap();
```

Comment:
Since we don't have support for opening DBs in read-only yet, please make sure that at least a `metadata.bin` file exists at the specified location, otherwise we just initialize an empty database here.